In this notebook we will demonstrate Auto ML on the mnist digits dataset. More information about autosklearn and autokeras can we found in the official documents [here](https://automl.github.io/auto-sklearn/master) and [here](https://github.com/keras-team/autokeras) respectively.

In [18]:
!sudo apt-get install build-essential swig
!curl https://raw.githubusercontent.com/automl/auto-sklearn/master/requirements.txt | xargs -n 1 -L 1 pip install
!pip install auto-sklearn

for _ in range(3):
    try:
        import autosklearn.classification
        break
    except:
        pass
else:
    raise ImportError("failed to import from autosklearn")

Reading package lists... Done
Building dependency tree       
Reading state information... Done
build-essential is already the newest version (12.4ubuntu1).
Suggested packages:
  swig-doc swig-examples swig3.0-examples swig3.0-doc
The following NEW packages will be installed:
  swig swig3.0
0 upgraded, 2 newly installed, 0 to remove and 31 not upgraded.
Need to get 1,100 kB of archives.
After this operation, 5,822 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 swig3.0 amd64 3.0.12-1 [1,094 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 swig amd64 3.0.12-1 [6,460 B]
Fetched 1,100 kB in 0s (9,015 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 2.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (T

### **Auto-Sklearn** on mnist dataset

In [0]:
import autosklearn.classification

import sklearn.datasets
import sklearn.metrics

import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split

In [2]:
#load the data
digits = sklearn.datasets.load_digits()
X = digits.data
y = digits.target
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, random_state=1)

#initialize the classifier
automl = autosklearn.classification.AutoSklearnClassifier()
automl.fit(X_train, y_train)
y_pred = automl.predict(X_test)

#accuracy
print("Accuracy score", sklearn.metrics.accuracy_score(y_test, y_pred))


Accuracy score 0.9888888888888889


We achived an accuracy of 98.9%.<br>
Now lets understand what exactly happened there.

In [3]:
print("Model\n",automl.show_models())
print("-")
print("Model Statistics\n",automl.sprint_statistics())

Model
 [(0.080000, SimpleClassificationPipeline({'balancing:strategy': 'none', 'classifier:__choice__': 'libsvm_svc', 'data_preprocessing:categorical_transformer:categorical_encoding:__choice__': 'one_hot_encoding', 'data_preprocessing:categorical_transformer:category_coalescence:__choice__': 'no_coalescense', 'data_preprocessing:numerical_transformer:imputation:strategy': 'mean', 'data_preprocessing:numerical_transformer:rescaling:__choice__': 'robust_scaler', 'feature_preprocessor:__choice__': 'liblinear_svc_preprocessor', 'classifier:libsvm_svc:C': 258.9142997145647, 'classifier:libsvm_svc:gamma': 0.008836682811147231, 'classifier:libsvm_svc:kernel': 'poly', 'classifier:libsvm_svc:max_iter': -1, 'classifier:libsvm_svc:shrinking': 'False', 'classifier:libsvm_svc:tol': 0.00017468203025368254, 'data_preprocessing:numerical_transformer:rescaling:robust_scaler:q_max': 0.7800242369785717, 'data_preprocessing:numerical_transformer:rescaling:robust_scaler:q_min': 0.25089878093468937, 'featu

### **Auto-Keras** on mnist dataset

In [19]:
#download the necessary packages you might have to resrart your runtime after this 
!pip install autokeras==1.0.2
!pip install tensorflow==2.1.0
#!pip install git+git://github.com/keras-team/keras-tuner.git@d2d69cba21a0b482a85ce2a38893e2322e139c01

     |████████████████████████████████| 421.8MB 26kB/s 
     |████████████████████████████████| 450kB 36.3MB/s 
     |████████████████████████████████| 3.9MB 23.9MB/s 
  Created wheel for gast: filename=gast-0.2.2-cp36-none-any.whl size=7540 sha256=064f68ae656f6e8f23f18ec601b2972d33654a60183d89d2aaab680b7cd53608
  Stored in directory: /root/.cache/pip/wheels/5c/2e/7e/a1d4d4fcebe6c381f378ce7743a3ced3699feb89bcfbdadadd
Successfully built gast
ERROR: tensorflow-probability 0.10.0 has requirement gast>=0.3.2, but you'll have gast 0.2.2 which is incompatible.
  Found existing installation: gast 0.3.3
    Uninstalling gast-0.3.3:
      Successfully uninstalled gast-0.3.3
  Found existing installation: tensorflow-estimator 2.2.0
    Uninstalling tensorflow-estimator-2.2.0:
      Successfully uninstalled tensorflow-estimator-2.2.0
  Found existing installation: tensorboard 2.2.1
    Uninstalling tensorboard-2.2.1:
      Successfully uninstalled tensorboard-2.2.1
  Found existing installation: 

In [0]:
from tensorflow.keras.datasets import mnist
import autokeras as ak
import warnings
warnings.filterwarnings('ignore')


In [11]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
print(x_train.shape)
print(y_train.shape)
print(y_train[:3]) 

(60000, 28, 28)
(60000,)
[5 0 4]


In [0]:
# split=2000

# x_test = x_test[:split]
# y_test = y_test[:split]

# x_train = x_train[:split]
# y_train = y_train[:split]


In [12]:
# Initialize the image classifier.
clf = ak.ImageClassifier(max_trials=10) # It tries 10 different models.
# Feed the image classifier with training data.
clf.fit(x_train,y_train, epochs=3)

Train for 1500 steps, validate for 375 steps
Epoch 1/3
1500/1500 [==============================] - 10s 7ms/step - loss: 0.1740 - accuracy: 0.9471 - val_loss: 0.0632 - val_accuracy: 0.9808
Epoch 2/3
1500/1500 [==============================] - 9s 6ms/step - loss: 0.0785 - accuracy: 0.9750 - val_loss: 0.0517 - val_accuracy: 0.9852
Epoch 3/3
1500/1500 [==============================] - 10s 6ms/step - loss: 0.0635 - accuracy: 0.9802 - val_loss: 0.0463 - val_accuracy: 0.9870


Train for 1500 steps, validate for 375 steps
Epoch 1/3
1500/1500 [==============================] - 110s 74ms/step - loss: 0.2727 - accuracy: 0.9259 - val_loss: 0.2531 - val_accuracy: 0.9405
Epoch 2/3
1500/1500 [==============================] - 105s 70ms/step - loss: 0.1165 - accuracy: 0.9701 - val_loss: 0.1070 - val_accuracy: 0.9706
Epoch 3/3
1500/1500 [==============================] - 105s 70ms/step - loss: 0.1016 - accuracy: 0.9733 - val_loss: 0.0666 - val_accuracy: 0.9803


Train for 1500 steps, validate for 375 steps
Epoch 1/3
1500/1500 [==============================] - 10s 7ms/step - loss: 0.1772 - accuracy: 0.9460 - val_loss: 0.0647 - val_accuracy: 0.9817
Epoch 2/3
1500/1500 [==============================] - 9s 6ms/step - loss: 0.0770 - accuracy: 0.9753 - val_loss: 0.0542 - val_accuracy: 0.9853
Epoch 3/3
1500/1500 [==============================] - 9s 6ms/step - loss: 0.0627 - accuracy: 0.9802 - val_loss: 0.0501 - val_accuracy: 0.9857


INFO:tensorflow:Oracle triggered exit
Train for 1875 steps, validate for 375 steps
Epoch 1/3
1875/1875 [==============================] - 12s 6ms/step - loss: 0.1570 - accuracy: 0.9519 - val_loss: 0.0474 - val_accuracy: 0.9851
Epoch 2/3
1875/1875 [==============================] - 12s 6ms/step - loss: 0.0736 - accuracy: 0.9775 - val_loss: 0.0267 - val_accuracy: 0.9926
Epoch 3/3
1875/1875 [==============================] - 12s 6ms/step - loss: 0.0602 - accuracy: 0.9814 - val_loss: 0.0198 - val_accuracy: 0.9948


In [13]:
#predict label
predicted_y = clf.predict(x_test)
print(predicted_y)  

[[7]
 [2]
 [1]
 ...
 [4]
 [5]
 [6]]


In [14]:
#evaluate on test data
print(clf.evaluate(x_test, y_test))

313/313 [==============================] - 1s 4ms/step - loss: 0.0363 - accuracy: 0.9877
[0.03626436673476883, 0.9877]
